uses flux for deleting delegates.

tdowds opened the pull request:
Using this PR to use flux for data updates (#523).

For this particular commit it handles deleting a delegate (#518). I'd like to get a review on this before implementing other updates. Furthermore I should probably update the DelegateStore test to reflect this change.

kmeht commented on huxley/www/js/actions/DelegateActions.js:
```diff
@@ -0,0 +1,37 @@
+  deleteDelegate(delegate) {
+    Dispatcher.dispatch({
+      actionType: ActionConstants.DELETE_DELEGATE,
+      delegate: delegate
```
We should only need a `delegateID` to delete a delegate, not the entire object.

kmeht commented on huxley/www/js/actions/DelegateActions.js:
```diff
@@ -0,0 +1,37 @@
+    Dispatcher.dispatch({
+      actionType: ActionConstants.UPDATE_DELEGATE,
+      delegateID: delegateID,
+      name: name,
```
Why are all the fields spread out here, vs. the single object in the add method? We should be consistent; I vote for the object here.

kmeht commented on huxley/www/js/components/AdvisorRosterView.js:
```diff
@@ -57,6 +58,18 @@ var AdvisorRosterView = React.createClass({
+        loading: false,
+        modal_open: false
+      });
+    }.bind(this));
```
Use an arrow function so we don't have to explicitly bind.
```js
DelegateStore.addListener(() => {
  //...
});
```

kmeht commented on huxley/www/js/components/AdvisorRosterView.js:
```diff
@@ -57,6 +58,18 @@ var AdvisorRosterView = React.createClass({
+      var delegates = DelegateStore.getDelegates(schoolID);
+      this.setState({
+        delegates: delegates,
+        loading: false,
```
Don't think we need to set either loading or modal_open in this callback. What was the reason for including them?

kmeht commented on huxley/www/js/stores/__tests__/DelegateStore-test.js:
```diff
@@ -8,18 +8,20 @@
 jest.dontMock('stores/DelegateStore');
 
 describe('DelegateStore', () => {
+  var ActionConstants
```
semicolon

kmeht commented on huxley/www/js/stores/__tests__/DelegateStore-test.js:
```diff
@@ -40,4 +42,41 @@ describe('DelegateStore', () => {
   });
+
+  it('adds a delegate', () => {
+    return Promise.all([
```
Don't need `Promise.all` if there's only one promise

kmeht commented on huxley/www/js/stores/__tests__/DelegateStore-test.js:
```diff
@@ -40,4 +42,41 @@ describe('DelegateStore', () => {
+    return Promise.all([
+      DelegateStore.getDelegates(mockSchoolId, (delegates) => {
+        expect(delegates).toEqual(mockDelegates);
+        var updatedDelegate = mockDelegates[0]
```
I don't think this test works as expected, because `delegates` and `mockDelegates` are literally the same object. That means, if you update one delegate's name in-place (like below), it already updates it in the store (which is no bueno).

Instead, just write out the new name in the `updateDelegate` call:
```js
DelegateStore.updateDelegate(
  d.id,
  'Jake Moskowitz',
  d.email,
  mockSchoolId,
);
```

kmeht commented on huxley/www/js/actions/DelegateActions.js:
```diff
@@ -0,0 +1,37 @@
+      delegateID: delegateID,
+      name: name,
+      email: email,
+      schoolID: schoolID
```
I don't think schoolID should be an updateable field. Opens us up to bugs where delegates are in the wrong school. we should also make this immutable on the server side, if we haven't already.

kmeht commented on huxley/www/js/stores/DelegateStore.js:
```diff
@@ -1,32 +1,79 @@
+
+  deleteDelegate(delegate) {
+    ServerAPI.deleteDelegate(delegate.id);
+    var index = _delegates.indexOf(delegate);
```
Assuming that we'd only pass the `delegateID` into this function:
```js
_delegates = _delegates.filter(d => d.id !== delegateID);
```

This sets `_delegates` to be a new array instead of mutating the old one. In general, we want to avoid mutative behaviors like this, because we don't know who else has a reference to these objects.

tdowds commented:
Updated this PR to do the following:
1. Does not depend on Promise in the Delegate Store
2. Updated Delegate Store methods to reflect changes in (1)
3. Updated AdvisorAssignmentView to reflect changes in 1 and 2 (Added a bulk update action on delegates for this)
4. Updated Delegate Store tests accordingly

Things still to do:
1. Use flux in AssignmentStore
2. Fire actions on successful or unsuccessful ServerAPI requests and update UI accordingly
For these two items I think I will open in separate diffs after this one is good to go

kmeht commented on huxley/www/js/components/AdvisorRosterView.js:
```diff
@@ -50,13 +51,23 @@ var AdvisorRosterView = React.createClass({
-    DelegateStore.getDelegates(user.school.id, function(delegates) {
-      this.setState({delegates: delegates});
-    }.bind(this));
+    this.setState({delegates: DelegateStore.getDelegates(user.school.id)});
```
This now belongs in `getInitialState`

kmeht commented on huxley/www/js/stores/DelegateStore.js:
```diff
@@ -1,32 +1,94 @@
+      email: delta.email,
+    });
+    var delegate = _delegates[delegateID];
+    delegate.name = delta.name;
```
This is still a mutative action. Instead, I'd use the spread operator to copy the delegate:
```js
const delegate  = {
  ..._delegates[delegateID],
  name,
  email,
};
_delegates[delegateID] = delegate;
_schoolsDelegates[delegate.school] = /*...*/;
```

kmeht commented on huxley/www/js/stores/DelegateStore.js:
```diff
@@ -1,32 +1,94 @@
+
+  addDelegate(delegate) {
+    _delegates[delegate.id] = delegate;
+    _schoolsDelegates[delegate.school].push(delegate);
```
This is a mutation, I'd use `concat` instead

kmeht commented on huxley/www/js/stores/DelegateStore.js:
```diff
@@ -1,32 +1,94 @@
+    delegate.name = delta.name;
+    delegate.email = delta.email;
+    _delegates[delegateID] = delegate;
+    _schoolsDelegates[delegate.school] = _schoolsDelegates[delegate.school].map(d => d.id == delegate.id ? delegate : d);
```
Same style nit, put the function on a separate line

kmeht commented on huxley/www/js/stores/DelegateStore.js:
```diff
@@ -1,32 +1,94 @@
+      schoolID,
+      JSON.stringify(delegates)
+    )
+    for (var i = 0; i < delegates.length; i++) {
```
Use a `for-of` loop instead:
```js
for (const delegate of delegates) {
  //...
}
```

kmeht commented on huxley/www/js/stores/DelegateStore.js:
```diff
@@ -1,32 +1,94 @@
+
+    ServerAPI.getDelegates(schoolID).then(value => {
+      _schoolsDelegates[schoolID] = value;
+      for (var i = 0; i < value.length; i++) {
```
Use a `for-of` loop.

kmeht commented on huxley/www/js/components/AdvisorAssignmentsView.js:
```diff
@@ -63,7 +64,18 @@ var AdvisorAssignmentsView = React.createClass({
       this.setState({countries: new_countries})
     }.bind(this));
-    DelegateStore.getDelegates(user.school.id, (delegates) => {
+    var delegates = DelegateStore.getDelegates(user.school.id);
```
Put the following in `getInitialState`
